### PR TITLE
8291286: Improve JavaDocs for static constructors in MemorySession

### PR DESCRIPTION
**Pull request by minborg**
This PR adds information to the static constructor methods in MemorySession to improve clarity for new users.

This is my first PR for a JDK project so I am eager to receive feedback and improvement suggestions from the community.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8291286](https://bugs.openjdk.org/browse/JDK-8291286): Improve JavaDocs for static constructors in MemorySession

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/693/head:pull/693` \
`$ git checkout pull/693`

Update a local copy of the PR: \
`$ git checkout pull/693` \
`$ git pull https://git.openjdk.org/panama-foreign pull/693/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 693`

View PR using the GUI difftool: \
`$ git pr show -t 693`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/693.diff">https://git.openjdk.org/panama-foreign/pull/693.diff</a>

</details>

### CODE REVIEW

**Comment by minborg** [comment]
With respect to OCA verification: I am an Oracle employee.

File: src/java.base/share/classes/java/lang/foreign/MemorySession.java [hunk 1]
```diff
@@ -228,56 +228,85 @@ default MemorySegment allocate(long bytesSize, long bytesAlignment) {
     }
 
     /**
-     * Creates a closeable confined memory session.
-     * @return a new closeable confined memory session.
+     * Creates and returns a new closeable, thread-confined memory session.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySession.java:231]: This `Creates and returns` is correct technically, but I can't find much evidence for it in the rest of the JDK. It gives 109 matches in current code base, whereas `Creates` has over 7K. I would leave this as is.

File: src/java.base/share/classes/java/lang/foreign/MemorySession.java [hunk 2]
```diff
@@ -228,56 +228,85 @@ default MemorySegment allocate(long bytesSize, long bytesAlignment) {
...
+     * <p>
+     * The returned memory session is confined to the current thread.
+     * <p>
+     * The returned memory session must eventually be closed to prevent memory leaks.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySession.java:235]: I suggest adding a linkplain on `closed` to point to the close method.

File: src/java.base/share/classes/java/lang/foreign/MemorySession.java [hunk 3]
```diff
@@ -228,56 +228,85 @@ default MemorySegment allocate(long bytesSize, long bytesAlignment) {
...
-     * Equivalent to (but likely more efficient than) the following code:
+     * Creates and returns a new non-closeable memory session that can be shared across threads,
+     * manage by a private {@link Cleaner} instance.
+     * <p>
+     * This is equivalent to (but likely more efficient than) the following code:
      * {@snippet lang=java :
      * openShared(Cleaner.create()).asNonCloseable();
      * }
-     * @return a non-closeable shared memory session, managed by a private {@link Cleaner} instance.
+     * @return a non-closeable, thread-shared memory session, managed by a private {@link Cleaner} instance
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/foreign/MemorySession.java:295]: Actually I would borrow a bit from what you did for `openShared`:

```
@return a new closeable memory session that can be shared across threads
```

so:

```
@return a new non-closeable memory session that can be shared across threads, managed by ...
```

File: src/java.base/share/classes/java/lang/foreign/MemorySession.java [hunk 4]
```diff
@@ -228,56 +228,85 @@ default MemorySegment allocate(long bytesSize, long bytesAlignment) {
...
         return MemorySessionImpl.createImplicit();
     }
 
     /**
      * Returns the global memory session.
-     * @return the global memory session.
+     * <p>
+     * The global memory session is non-closeable and can be shared across threads.
+     * <p>
+     * The global memory session is neither ever closed nor ever subject to cleaning.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySession.java:306]: These comments seem out of place, in the sense that the concept of memory session is defined in the top level javadoc. I'm not sure repeating what a global session is here adds much? If we really think that the old text is not good enough, then maybe pulling out the global session text into a subsection in the toplevel javadoc, and then referring to that one from here could be a better option (although, again, not sure this needs changing?).

File: src/java.base/share/classes/java/lang/foreign/MemorySession.java [hunk 5]
```diff
@@ -228,56 +228,84 @@ default MemorySegment allocate(long bytesSize, long bytesAlignment) {
...
      */
     static MemorySession openImplicit() {
         return MemorySessionImpl.createImplicit();
     }
 
     /**
      * Returns the global memory session.
-     * @return the global memory session.
+     *
+     * @return the global memory session
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySession.java:308]: Use {@return} tag please, to avoid duplication

**Review by mcimadamore — APPROVED** [review]

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**